### ISSUE TITLE
Extension types: Updates regarding covariance and `UP`

### ISSUE THREAD

**Issue by eernstg** [dart-lang/co19#2217]
If and when https://github.com/dart-lang/language/pull/3276 lands, we need additional tests:

If `E` is a generic extension type and `S1 <: T1` .. `Sk <: Tk` then `E<S1 .. Sk>` is a subtype of `E<T1 .. tk>`.

It also specifies that the Dart 1 algorithm for computation of the least upper bound of two types (the one which is used in the last case of the function UP, [here](https://github.com/dart-lang/language/blob/main/resources/type-system/upper-lower-bounds.md#:~:text=UP(C0%3CT0%2C%20...%2C%20Tn%3E%2C%20C1%3CS0%2C%20...%2C%20Sk%3E)%20%3D)), must take `Object?` into account:

```dart
extension type E1(Object? o) {} // Depth 1
extension type E2(Object? o) implements E1 {} // Depth 2
extension type E3a(num n) implements E2, num {} // Depth 3
extension type E3b(num n) implements E2, num {} // Depth 3
```

In this case, with the old rule, `E1` and `E2` would not have a well-defined depth, but `num` had depth 1. `E3a` and `E3b` could have depth 2, or it could be undefined because some superinterface paths do not reach `Object`.

We need to consider the depth from `Object?` rather than `Object`, so all the "old" types are now 1 step deeper (`num` now has depth 2), and every type has a well-defined depth (in particular, `E3a` and `E3b` have depth 3).

For `UP(E3a, E3b)` we get `Object?`, because `E2` and `num` both have depth 2.

**Comment by eernstg** [comment]
(Looks like we have covariance tests already in https://github.com/dart-lang/co19/pull/2215, but that shouldn't hurt. The landed version of dart-lang/language#3276 will _surely_ include covariance.)

**Comment by sgrekhov** [comment]
Fixed by #2217
